wandb/wandb
# feat(launch): Launch agent uses thread pool to run jobs

Fixes [WB-12303](https://wandb.atlassian.net/browse/WB-12303)

Description
-----------
The launch agent has been refactored to use a thread pool, so that when a run hangs during creation, other runs can still be started.

A side effect of this change is that we no longer kill the subprocess for a local submitted run, see comments for details.

Testing
-------
- Run launch agent with max-threads=2
- Run `wandb launch ...` multiple times, queuing several jobs
- Confirm that the launch agent is capable of running multiple jobs and the logging correctly reflects the agent state


[WB-12303]: https://wandb.atlassian.net/browse/WB-12303?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
